trieste/utils/pareto.py

```diff
@@ -0,0 +1,309 @@
+        :param datasets: Mapping of Dataset and its coresponding string key(s)
+        :param generic_strategy: Force the generic divide and conquer strategy
+            regardless of the number of objectives (default False)
+        """
+
+        if datasets is not None:
+            data = to_Dataset(datasets)
+        self.data = data if datasets is not None else self.data
+
+        # Find (new) set of non-dominated points
+        changed = self._update_front()
+
+        # Recompute cell bounds if required
+        # Note: if the Pareto set is based on model predictions
+        # it will almost always change in between optimizations
+        if changed:
+            # Clear data container
+            self.bounds.clear()
+            if generic_strategy:
+                self.divide_conquer_nd()
+            else:
+                self.bounds_2d() if self.data.observations.shape[
+                    1
+                ] == 2 else self.divide_conquer_nd()
+
+    def divide_conquer_nd(self):
+        """
+        Divide and conquer strategy to compute the cells covering the non-dominated region.
+        Generic version: works for an arbitrary number of objectives.
+        """
+        outdim = self.data.observations.shape[1]
+
+        # The divide and conquer algorithm operates on a pseudo Pareto set
+        # that is a mapping of the real Pareto set to discrete values
+        pseudo_pf = (
+            tf.argsort(self.front, axis=0) + 1
+        )  # +1 as index zero is reserved for the ideal point
+
+        # Extend front with the ideal and anti-ideal point
+        min_pf = tf.reduce_min(self.front, axis=0) - 1
+        max_pf = tf.reduce_max(self.front, axis=0) + 1
+
+        pf_ext = tf.concat(
+            [atleast2d(min_pf), atleast2d(self.front), atleast2d(max_pf)], axis=0
+        )  # Needed for early stopping check (threshold)
+        pf_ext_idx = tf.concat(
+            [
+                atleast2d(tf.zeros(outdim, dtype=tf.int32)),
+                atleast2d(pseudo_pf),
+                atleast2d(tf.ones(outdim, dtype=tf.int32) * self.front.shape[0] + 1),
+            ],
+            axis=0,
+        )
+
+        # Start with one cell covering the whole front
+        dc = [
+            (
+                tf.zeros(outdim, dtype=tf.int32),
+                (int(pf_ext_idx.shape[0]) - 1) * tf.ones(outdim, dtype=tf.int32),
+            )
+        ]
+        total_size = tf.reduce_prod(max_pf - min_pf)
+
+        # Start divide and conquer until we processed all cells
+        while dc:
+            # Process test cell
+            cell = dc.pop()
+            arr = tf.range(outdim)
+            idx_lb = tf.gather_nd(pf_ext_idx, tf.stack((cell[0], arr), -1))
+            idx_ub = tf.gather_nd(pf_ext_idx, tf.stack((cell[1], arr), -1))
+            lb = tf.gather_nd(pf_ext, tf.stack((idx_lb, arr), -1))
+            ub = tf.gather_nd(pf_ext, tf.stack((idx_ub, arr), -1))
```

Comment:
a general note: `gather_nd` is slow, and you've used it liberally in this PR. If you're able to find faster functions, that'd be good